File: shortest-distance-to-a-character/shortest_distance_to_a_character.go
func shortestToChar(S string, C byte) []int {
	indexes := make([]int, 0)
	for i, c := range S {
		if byte(c) == C {
			indexes = append(indexes, i)
		}
	}

	result := make([]int, len(S))
	index, ind, stop := 0, 0, 0
	for index = 1; index < len(indexes); index++ {
		for ind = stop; ind < len(S); ind++ {
			if ind == indexes[index] {
				result[ind] = 0
				stop = ind + 1
				break
			} else if ind <= indexes[index-1] {
				result[ind] = indexes[index-1] - ind
			} else if ind > indexes[index-1] && ind < indexes[index] {
				result[ind] = min(ind-indexes[index-1], indexes[index]-ind)
			}
		}
	}

	for len(S) > ind {
		result[ind] = max(ind-indexes[index-1], indexes[index-1]-ind)
		ind++
	}

	return result
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func min(a, b int) int {
	if a > b {
		return b
	}
	return a
}

File: shortest-distance-to-a-character/shortest_distance_to_a_character.py
class Solution:
    def shortestToChar(self, S: str, C: str) -> List[int]:
        indexes = [i for i in range(len(S)) if S[i] == C]
        result = [float('inf') for c in S]
        index, i, stop = 0, 0, 0

        for index in range(1, len(indexes)):
            for i in range(stop, len(S)):
                if i == indexes[index]:
                    result[i] = 0
                    stop = i+1
                    break
                elif i <= indexes[index-1]:
                    result[i] = indexes[index-1]-i
                elif i > indexes[index-1] and i < indexes[index]:
                    result[i] = min(i-indexes[index-1], indexes[index]-i)

        while len(S) > i:
            result[i] = max(i-indexes[index], indexes[index]-i)
            i += 1

        return result
